### PR TITLE
monero-wallet-cli: Stop refresh when top block is reached #1190

### PR DESCRIPTION
**Pull request by revler1082**
Clamps user supplied restore height to daemon blockchain height.

### CODE REVIEW

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -1,21 +1,21 @@
 // Copyright (c) 2014-2016, The Monero Project
-// 
+//
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: No whitespace spam changes please

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -1206,8 +1206,22 @@ void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& re
...
   m_run.store(true, std::memory_order_relaxed);
   if (start_height > m_blockchain.size() || m_refresh_from_block_height > m_blockchain.size()) {
-    if (!start_height)
+    if (!start_height) {
+
+      std::string daemon_height_err = "";
+      uint64_t daemon_bc_height = this->get_daemon_blockchain_height(daemon_height_err);
+      if(daemon_height_err.size() > 0) {
+        throw std::runtime_error(daemon_height_err);
+      } else if(m_refresh_from_block_height > daemon_bc_height) {
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: Should be >=, as the first is a block height and the second is a blockchain height (it can be a bit confusing).

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -1206,8 +1206,22 @@ void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& re
   get_short_chain_history(short_chain_history);
   m_run.store(true, std::memory_order_relaxed);
   if (start_height > m_blockchain.size() || m_refresh_from_block_height > m_blockchain.size()) {
-    if (!start_height)
+    if (!start_height) {
+
+      std::string daemon_height_err = "";
+      uint64_t daemon_bc_height = this->get_daemon_blockchain_height(daemon_height_err);
+      if(daemon_height_err.size() > 0) {
+        throw std::runtime_error(daemon_height_err);
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: That should throw another exception, so catch blocks match the actual error. But then it's probably due to get_daemon_blockchain_height not throwing the right one in the first place so I get where you're coming from :)

**Comment by moneromooo-monero** [comment]
Makes sense, but things need fixing first.

File: src/wallet/wallet2.cpp [hunk 4]
```diff
@@ -1206,8 +1206,22 @@ void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& re
   get_short_chain_history(short_chain_history);
   m_run.store(true, std::memory_order_relaxed);
   if (start_height > m_blockchain.size() || m_refresh_from_block_height > m_blockchain.size()) {
-    if (!start_height)
+    if (!start_height) {
+
+      std::string daemon_height_err = "";
+      uint64_t daemon_bc_height = this->get_daemon_blockchain_height(daemon_height_err);
```
Comment (Jaqueeee) [inline src/wallet/wallet2.cpp]: What happens when the daemon isn't fully synced? It looks like it would set m_refresh_from_block_height to current daemon height, abort fast_refresh() on that block and continue refresh with the slower pull_blocks(). Not sure why we would wan't that? 
Maybe wallet2::get_daemon_blockchain_target_height() could be used instead?

Comment (revler1082) [inline src/wallet/wallet2.cpp]: Thanks for pointing out the pitfalls. I attempted to address these in the latest commit. Do you think it's better to just set the restore height to target if it's way high, or actually stick to the user specified height? The latest commit just goes with user height no matter how far in the future.

Comment (NanoAkron) [inline src/wallet/wallet2.cpp]: Probably safest to inform the user that their value exceeds local maximum block height, because I can't see what value setting the restore point in the future would have if it's an action they want to run now.

Comment (Jaqueeee) [inline src/wallet/wallet2.cpp]: I agree. Better to just inform the user. 
It's kind of hard to follow your changes now because of a lot of whitespace changes.  
I think you should check your editor settings or commit/add using --ignore-whitespace to make the commits cleaner.

Comment (revler1082) [inline src/wallet/wallet2.cpp]: Yep, sorry about all of the whitespace, i'll start using that flag.

Comment (Jaqueeee) [inline src/wallet/wallet2.cpp]: Fyi: Just found out that you could append ?w=1 to github urls to view without whitespace.  
https://github.com/monero-project/monero/pull/1197/files?w=1

**Comment by moneromooo-monero** [comment]
Did you update the patch ? There's a system comment from github implying you did, but the diffs look the same as last time, so I'm not sure...

**Comment by revler1082** [comment]
Hmm, fluffy recently merged one of my other branches (coinbase-tx-sum), could that be the confusion?

**Comment by danrmiller** [comment]
The compile is failing for osx and for 32 bit windows and 32 bit linux. I know its a little hard to notice right now in the buildbot status that gets reported to github because of the tests, so I wanted to point it out here, since something got merged before. You can click above for details.

osx:
`[ 96%] Linking CXX executable ../../bin/monerod
Undefined symbols for architecture x86_64:
  "bool cryptonote::Blockchain::get_blocks<unsigned long long, unsigned long long const, std::__1::list<cryptonote::block, std::__1::allocator<cryptonote::block> > >(unsigned long long const&, unsigned long long const&, std::__1::list<cryptonote::block, std::__1::allocator<cryptonote::block> >&) const", referenced from:
      cryptonote::core::get_coinbase_tx_sum(unsigned long long, unsigned long long) in libcryptonote_core.a(cryptonote_core.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [bin/monerod] Error 1
make[2]: *** [src/daemon/CMakeFiles/daemon.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [release-static-64] Error 2`

linux i686:
`[ 95%] Linking CXX executable ../../bin/monerod
/home/vagrant/slave/monero-static-ubuntu-i686/build/src/daemon/daemon.cpp: In member function 'run':
/usr/include/c++/5/bits/atomic_base.h:374:18: warning: 'rpc_commands' may be used uninitialized in this function [-Wmaybe-uninitialized]
  __atomic_store_n(&_M_i, __i, __m);
                  ^
/home/vagrant/slave/monero-static-ubuntu-i686/build/src/daemon/daemon.cpp:123:34: note: 'rpc_commands' was declared here
     daemonize::t_command_server* rpc_commands;
                                  ^
/tmp/ccARHk3x.ltrans5.ltrans.o: In function `cryptonote::core::get_coinbase_tx_sum(unsigned long long, unsigned long long)':
<artificial>:(.text+0x3667): undefined reference to `bool cryptonote::Blockchain::get_blocks<unsigned long long, unsigned long long const, std::__cxx11::list<cryptonote::block, std::allocator<cryptonote::block> > >(unsigned long long const&, unsigned long long const&, std::__cxx11::list<cryptonote::block, std::allocator<cryptonote::block> >&) const'
collect2: error: ld returned 1 exit status
src/daemon/CMakeFiles/daemon.dir/build.make:262: recipe for target 'bin/monerod' failed
make[3]: Leaving directory '/home/vagrant/slave/monero-static-ubuntu-i686/build/build/release'
make[3]: *** [bin/monerod] Error 1`

**Comment by revler1082** [comment]
@danrmiller thanks for pointing it out, sorry for being such a noob O.o

This is probably a silly question, but if I make the patch on my coinbase-tx-sum branch, should I open another pull request?

**Comment by moneromooo-monero** [comment]
One overarching problem: loads of irrelevant commits from other people in that PR. This makes it hard to review, adds plenty of spam in changes. You probably merged, rather than rebased.
To rebase:
- make sure your master branch is up to date
- move to your topic branch
- git rebase master
Your branch will not be clean.

I'll comment on the diff now.

**Comment by moneromooo-monero** [comment]
And there's still loads of spam whitespace changes.

**Comment by moneromooo-monero** [comment]
Good practice is also to squash fixup commits, so you don't actually push a wrong commit.

**Comment by revler1082** [comment]
@moneromooo-monero ugh apologies for the headache, to say my git skills are sub-par would be an understatement. I was trying to rebase this branch so it would be easier to merge, but I obviously messed that up.

**Comment by NanoAkron** [comment]
This stuff certainly improves your git skills quickly :)

**Comment by moneromooo-monero** [comment]
Sorry, I'd forgot about this. The commits look all rebased fine now, thanks. I'll review in a few hours.

File: src/wallet/wallet2.cpp [hunk 5]
```diff
@@ -1202,12 +1202,27 @@ void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& re
...
-    if (!start_height)
-      start_height = m_refresh_from_block_height;
+    
+    std::string daemon_target_err = "";
+    uint64_t daemon_target_height = this->get_daemon_longest_known_chain_height(daemon_target_err);
+    if(daemon_target_err.size() > 0) {
+      throw std::runtime_error(daemon_target_err);
+    }    
+    
+    if (m_refresh_from_block_height > daemon_target_height) m_refresh_from_block_height = daemon_target_height;
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: That seems wrong. daemon_target_height would be a blockchain height, so you'd likely need m_refresh_from_block_height = daemon_target_height - 1;

File: src/wallet/wallet2.cpp [hunk 6]
```diff
@@ -1202,12 +1202,27 @@ void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& re
   std::list<cryptonote::block_complete_entry> blocks;
   std::vector<COMMAND_RPC_GET_BLOCKS_FAST::block_output_indices> o_indices;
 
+  std::string daemon_height_err = "";
+  uint64_t daemon_bc_height = this->get_daemon_blockchain_height(daemon_height_err);
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: No need for the this-> constructs in the typical case.

File: src/wallet/wallet2.cpp [hunk 7]
```diff
@@ -4108,6 +4126,38 @@ uint64_t wallet2::get_daemon_blockchain_target_height(string &err)
...
+      err = "";
+    }
+  }
+  else
+  {
+    err = "possibly lost connection to daemon";
+  }
+  return resp_t.result.height;
+}
+
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: Does this give any other result than the existing get_daemon_blockchain_height ?

**Comment by moneromooo-monero** [comment]
A few comments, but generally ok. Please make sure to test it well, because I remember it was easy to break.

**Comment by moneromooo-monero** [comment]
ping! I think my comments went through, but let me know if you don't see them, I've kept them unsubmitted a few times before.

**Comment by revler1082** [comment]
Hi @moneromooo-monero , yep the comments are there, I've just been busy with work and some other stuff. Thanks again for all your help!

**Comment by fluffypony** [comment]
Just to update: I'm waiting on those changes before merging, but I am keeping my eye on it:)

**Comment by fluffypony** [comment]
Still waiting :)

**Comment by revler1082** [comment]
Apologies gents, I'm out at some work conference thing at the moment (goes thru the week), but I'll try to get everything patched up by this weekend.

**Comment by fluffypony** [comment]
@revler1082 don't forget about this :)

**Comment by revler1082** [comment]
I haven't tested this latest version using the approximate_height call, mostly pushing to work on at home, but feel free to take a quick glance as a sanity check. @moneromooo-monero was right about the duplicate height calls, so I just removed the longest_known_height function.
